### PR TITLE
Allow user to join an existing shopping list

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person
## Summary
We want to allow users to join existing shopping lists so they can manage them with friends or family. This would take the following steps:
1. User enters an existing token into a form
2. The token is saved to `localStorage`
3. The user is taken to the `List` view and shown the existing list.

## Acceptance criteria
If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Notes
- Be sure to use the `setListToken` function to save the token that the user enters!
- You will know your `<label>` is correctly implemented if you click on the label and keyboard focus moves to the related input. Refer to [the MDN docs on `<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) for more info.
- Users will be able to submit the form with the `Enter` key if you listen for the right JavaScript event on the right element! If you're adding an `onKeyDown` listener anywhere, you might be overcomplicating things.
- You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.
- If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 3).

### PR DESCRIPTION
**Pull request by jongranados**
## Description
This PR allows users to join existing shopping list so they can manage them with friends or family. This functionality takes the following steps:
1. The user enters a three word token existing token into a form on the home page.
2. The token is saved to `localStorage`.
3. The user is taken to the `List` view and shown the existing list.
4. If the list does not exist, the user sees an alert stating that the list does not exist.

The alert is generated here with Toastify which might present some accessibility concerns. The dev team will likely want to discuss the best route for generating alerts throughout the app.

## Related Issue
Closes #5.

## Acceptance Criteria
If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.
## Type of Changes
|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
| ✓    | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
This screenshot shows the application's home page before this functionality was implemented; there was no way for the user to join an existing shopping list unless an existing list's respective token was already stored in their browser's local storage.

before

### After
This screenshot show the application's home page after this functionality was implemented; a new user is given the option to either create a brand new list or join an existing list with the submission of a valid three word token. 

after

This video demonstrates a user attempting to access an existing list. They first attempt to access a list associated with token `this should fail`. Because such a token does not exist in the database, they are notified of their failed attempt through a pop-up message that reads: "Sorry, this list does not exist." The user then attempts to access a list with token `my test list`. Because this token does exist, the user is presented with the grocery items associated with this token.  

https://user-images.githubusercontent.com/9102330/233540840-763429d1-29d0-4976-96ba-370eb66ae51f.mov

## Testing Steps / QA Criteria
- Launch application: 
    - remotely
        - [preview URL for latest PR (97e1d05)](https://tcl-57-smart-shopping-list--pr21-es-jg-join-existing-1hai28xb.web.app/)
    - locally
        - Pull down `es-jg-join-existing-list`
        - Install dependencies with `npm i`
        - Start the application with `npm start`
- Clear your browser local storage of a pre-existing `tcl-shopping-list-token` 
- Navigate home (`/`)
- Attempting to submit an invalid token such as `this should fail` should:
    - keep you on the home page (`/`)
    - display a pop-up that reads: "Sorry, this list does not exist."
    - not result in a `tcl-shopping-list-token` inside your browser's local storage
- Attempting to submit a valid token such as `my test list` should:
    - redirect you to the list page (`/list`)
    - render the items in your list 
    - save the validated token inside your browser's local storage under `tcl-shopping-list-token`

### CODE REVIEW

File: src/api/firebase.js [hunk 1]
```diff
@@ -15,6 +15,16 @@ export function streamListItems(listId, handleSuccess) {
 	return onSnapshot(listCollectionRef, handleSuccess);
 }
 
+/**
+ * Check existence of list in Firestore associated with user token input.
+ * @param {string} userTokenInput The user's token input requesting to be validated
+ */
+export async function validateToken(userTokenInput) {
+	const listCollectionRef = collection(db, userTokenInput);
+	const listSnapshot = await getDocs(listCollectionRef);
```
Comment (Amy-Pr) [inline src/api/firebase.js:24]: Curious how line 23 and 24 work together? We looked up the firebase "getDocs" function. Still a little fuzzy on how it all works.

**Review by Amy-Pr — APPROVED** [review]
This looks super great guys! We tested it from the preview link. Works as expected. I even used a token name saved in Beth's browser, and was able to bring up her list on my end! Awesome job!!!

File: src/api/firebase.js (hunk 1 above)
Comment (jongranados) [reply src/api/firebase.js:24]: Hey Amy! If I'm understanding things correctly, accessing data from a firestore db is a two-part process. You first get a reference to a collection in the db using the `collection` method. I like to think of this like requesting an address. And with that reference, you're then able to actually query the db using `getDocs` to take "snapshots" of the db state at a particular instance in time. I like to think of this like a camera taking a picture of the db and capturing its state at an instance in time. 

Still hazy with all of this myself though so I'm not sure this is completely right.

File: src/api/firebase.js [hunk 2]
```diff
@@ -15,6 +15,16 @@ export function streamListItems(listId, handleSuccess) {
 	return onSnapshot(listCollectionRef, handleSuccess);
 }
 
+/**
+ * Check existence of list in Firestore associated with user token input.
```
Comment (djtaylor8) [inline src/api/firebase.js:19]: Love the documentation here!

File: src/views/Home.jsx [hunk 3]
```diff
@@ -2,8 +2,12 @@ import { Navigate } from 'react-router-dom';
 import './Home.css';
 import { generateToken } from '@the-collab-lab/shopping-list-utils';
 import * as React from 'react';
+import { validateToken } from '../api/firebase';
+import Toastify from 'toastify-js';
+import 'toastify-js/src/toastify.css';
 
 export function Home({ setListToken }) {
+	const [userTokenInput, setUserTokenInput] = React.useState('');
```
Comment (djtaylor8) [inline src/views/Home.jsx]: Suggesting again - import `{ useState } from 'react'` as opposed to `React.useState`. Not a blocker, but a bit cleaner IMO.

**Review by djtaylor8 — APPROVED** [review]
Great work, @jongranados and @emilysellers! I definitely think it is worth talking about the accessibility issues with everyone on the team with regards to using the Toastify library. Otherwise, the functionality is working as expected!

File: src/views/Home.jsx (hunk 3 above)
Comment (jongranados) [reply src/views/Home.jsx]: Whoops, forgot about this! Done. Definitely cleaner!

File: src/api/firebase.js (hunk 1 above)
Comment (mxmason) [reply src/api/firebase.js:24]: That's a great explanation!

File: src/api/firebase.js [hunk 4]
```diff
@@ -15,6 +15,16 @@ export function streamListItems(listId, handleSuccess) {
...
+ * @param {string} userTokenInput The user's token input requesting to be validated
+ */
+export async function validateToken(userTokenInput) {
+	const listCollectionRef = collection(db, userTokenInput);
+	const listSnapshot = await getDocs(listCollectionRef);
+	return !listSnapshot.empty;
+}
```
Comment (mxmason) [inline src/api/firebase.js:26]: This function works exactly as intended, but I have an optimization mission, should you choose to accept it.

Before we get to the question "does this list exist on the server?", we know our tokens have to be two things:
1. non-empty, and
2. a specific shape

Can we use this knowledge to check for validity and return early _without_ wasting time and resources on a request to our database?

File: src/views/Home.jsx [hunk 5]
```diff
@@ -1,25 +1,59 @@
...
+	}
+
 	return (
 		<div className="Home">
-			<p>
-				Hello from the home (<code>/</code>) page!
-			</p>
 			<button type="button" onClick={handleClick}>
-				Create New List
+				Create new list
```
Comment (mxmason) [inline src/views/Home.jsx:38]: Yesssss! I'm such a fan of removing unnecessary title case from our UIs!

**Review by mxmason — APPROVED** [review]
Great work! One optional idea for an optimization before we ship.

**Review by danainjax — APPROVED** [review]
Excellent job on this issue. I especially liked the explicit and complete pr, and the step by step instructions on how to test, and what you considered re: toastify and there may need to be discussion regarding tradeoffs or approach. Fantastic!

File: src/api/firebase.js (hunk 1 above)
Comment (bethmelmtv) [reply src/api/firebase.js:24]: great explanation- thank you @jongranados !

**Review by bethmelmtv — APPROVED** [review]
Great work @emilysellers  and @jongranados ! Everything's working :) Awesome job!

File: src/api/firebase.js (hunk 4 above)
Comment (jongranados) [reply src/api/firebase.js:26]: Agh, such a great point EJ! I'll incorporate that now.

Comment (jongranados) [reply src/api/firebase.js:26]: Actually, before doing so I want to ask a question at today's sync. Going to merge to main for now.

Comment (emilysellers) [reply src/api/firebase.js:26]: Great point, EJ! It would be great to talk this over today.

**Pull request merged**